gitpod-io/gitpod
# User Namespaced Workspaces

This PR wraps an entire workspace in a user namespace, which provides "root" within the workspace, while on the node they're never actually privileged.

### How does this work?
User namespaces shift the UID/GID of a user inside the namespace to different IDs outside the namespace. This way a user can have capabilities inside the namespace which aren't worth much outside the namespace.
To make this work we need to shift the workspace root filesystem to the mapped UID/GIDs. We do that using shiftfs. The shiftfs mark mount is performed by ws-daemon, upon request. Same with its un-mounting: it's unmounted by ws-daemon, triggered by supervisor.

Supervisor starts itself in a series of "rings", each of which successively adds new namespaces and mounts. The process looks like this:

- `ring0` creates a new user and mount namespace. starts ring1
- `ring1` connects to the ws-daemon socket at `/.workspace/daemon.sock`, triggers the UID/GID mapping writes, and makes ws-daemon mount the container's rootfs as shiftfs mark at `/.workspace/mark`. starts ring2
- `ring2` mounts the mark mount, thereby shifting the rootfs into the user namespace, makes a bunch of rbind mounts and chroots into the shifted rootfs. Starts `supervisor run`
- `supervisor run` runs until kubernetes SIGTERMs it
- `ring2` exits
- `ring1` asks ws-daemon to unmount the shiftfs mark, then exits
- `ring0` exits

One of the bigger changes to make all this work is that supervisor's `InWorkspaceHelper` and ws-daemon canaries were replaced with unix socket. In the past, ws-daemon would provide services to a workspace by connecting to the supervisor API endpoint and offering 'canaries'. The workspace could then talk to those canaries to trigger functionality from ws-daemon. This "inversion of control" was done to protect ws-daemon from potentially malicious workspaces, and because the network policy of workspaces (rightfully so) prevents them from connecting to Gitpod services other than server.
Now, for FWB and user-namespaced workspaces, ws-daemon places a unix socket in `<working-area>/<instanceID>-daemon`, which these workspaces mount as `hostPath`. The socket is serviced using workspace lifecycle hooks, and has rate limits on its RPC calls.

### How to test?
- make sure your user exists, i.e. visit http://cw-userns-ii.staging.gitpod-dev.com/workspaces/
- give yourself the `privileged-ws` permission and `user_namespace` feature flag by running:
  ```
  leeway run components/ws-daemon:enabled-userns-ff
  ```
- start [any workspace](http://cw-userns-ii.staging.gitpod-dev.com/#https://github.com/gitpod-io/django-locallibrary-tutorial)
- do something privileged (e.g. `sudo apt-get update && sudo apt-get install -y bb`)
- bonus points: validate we're still "just" UID 33333 on the node:
  - run `sudo sleep 1234` in a user namespaced workspace
  - verify you're not actually root
    ```bash
    WSPOD=<your-workspace-pod-name>
    WSPOD_NODE=$(kubectl get -o json pod $WSPOD | jq -r .spec.nodeName)
    WSPOD_DAEMON=$(kubectl describe node $WSPOD_NODE | grep userns-ii | grep ws-daemon | awk '{print $2}')
    kubectl exec -it $WSPOD_DAEMON -- ps guax | grep "sleep 1234"
    ```
- ensure the workspace backup still works (e.g. `touch /workspace/foo`, restart the workspace)

### Caveats, Open Points and Future Work
- unconfined seccomp: the containerd default seccomp profile seems not to allow `CLONE_USERNS`, which if I read [the sources](https://github.com/containerd/containerd/blob/ba8377590d326a5f73e4abe2d282c46c385329ae/contrib/seccomp/seccomp_default.go#L633-L647) correctly, it should. We have to clarify how we can come back to a more confined seccomp setup (either as supervisor ring, or preferably using the default runtime profile).
- sudo has [a "bug" where it complains](https://github.com/sudo-project/sudo/issues/42) about being unable to set the rlimit. The bug is fixed, but the ubuntu packages aren't updated yet. It's just a warning, sudo works just fine otherwise.
- the removal of the canaries broke the Git status for FWB workspaces (possibly FWB in general). Until we either fix FWB for minio or go back to GCP storage in devstaging, this is hard to debug.
- proc mounts: ideally we'd have ring2 create a PID namespace to hide all rings from the workspace workload. Due to the necessary [proc masking](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#allowedprocmounttypes), the [kernel won't let us mount](https://github.com/torvalds/linux/blob/22230cd2c55bd27ee2c3a3def97c0d5577a75b82/fs/namespace.c#L3966-L3978) `/proc`. We will probably work around this by delegating the proc mount to `ws-daemon`. However, it remains unclear how we'll work with this for running Docker in the workspace.
